### PR TITLE
chore(middleware-signing): refactor middleware to modular functions

### PR DESCRIPTION
**Pull request by trivikr**
### Issue
Refs: https://github.com/aws/aws-sdk-js-v3/issues/2208

### Description
Refactors middleware-signing to modular functions.

### Testing
CI

---
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/middleware-signing/src/middleware.ts [hunk 1]
```diff
@@ -10,23 +10,21 @@ import {
...
+import { getUpdatedSystemClockOffset } from "./utils/getUpdatedSystemClockOffset";
+
+export const awsAuthMiddleware =
+  <Input extends object, Output extends object>(
+    options: AwsAuthResolvedConfig
+  ): FinalizeRequestMiddleware<Input, Output> =>
+  (next: FinalizeHandler<Input, Output>, context: HandlerExecutionContext): FinalizeHandler<Input, Output> =>
     async function (args: FinalizeHandlerArguments<Input>): Promise<FinalizeHandlerOutput<Output>> {
       if (!HttpRequest.isInstance(args.request)) return next(args);
-      const signer = typeof options.signer === "function" ? await options.signer() : options.signer;
```
Comment (AllanZhengYP) [inline packages/middleware-signing/src/middleware.ts:25]: Why is this removed? Can you confirm why was it added?

Comment (AllanZhengYP) [reply packages/middleware-signing/src/middleware.ts:25]: Looks like it was added in https://github.com/aws/aws-sdk-js-v3/pull/472. Can you make sure we don't rely on it anymore?

File: packages/middleware-signing/src/middleware.spec.ts [hunk 2]
```diff
@@ -2,166 +2,91 @@ import { HttpRequest } from "@aws-sdk/protocol-http";
...
-    clockSkewNotPresent.forEach((systemClockOffsetVal) => {
-      it(`current systemClockOffset: ${systemClockOffsetVal.current}, new systemClockOffset: ${systemClockOffsetVal.new}`, async () => {
-        expect.assertions(3);
-        const systemClockOffset = systemClockOffsetVal.current;
-        const newSystemClockOffset = systemClockOffsetVal.new;
-        const options = {
-          signer: noOpSigner,
-          systemClockOffset,
-        };
+  describe("should update systemClockOffset if date header is present", () => {
```
Comment (AllanZhengYP) [inline packages/middleware-signing/src/middleware.spec.ts:75]: Add a case when date header does not exist?

File: packages/middleware-signing/src/middleware.ts (hunk 1 above)
Comment (trivikr) [reply packages/middleware-signing/src/middleware.ts:25]: The check for typeof to be function is removed as it's not required.

The signer is always a Provider in `AwsAuthResolvedConfig`
https://github.com/trivikr/aws-sdk-js-v3/blob/fac5f27fc2727d91be49e403a1a81005efe685d6/packages/middleware-signing/src/configurations.ts#L78-L95

The normalization happens in resolveAwsAuthConfig
 https://github.com/trivikr/aws-sdk-js-v3/blob/fac5f27fc2727d91be49e403a1a81005efe685d6/packages/middleware-signing/src/configurations.ts#L107-L110

File: packages/middleware-signing/src/middleware.spec.ts (hunk 2 above)
Comment (trivikr) [reply packages/middleware-signing/src/middleware.spec.ts:75]: The first unit test checks for absence of date header by verifying that `getUpdatedSystemClockOffset` is not called.

https://github.com/trivikr/aws-sdk-js-v3/blob/13fbdb3548e1d3fb5632157b0abf0d0483c40d5d/packages/middleware-signing/src/middleware.spec.ts#L52-L59

**Review by AllanZhengYP — APPROVED** [review]
